Sources/UIGestureRecognizer.swift

```diff
@@ -29,17 +29,19 @@ open class UIGestureRecognizer {
 
     public var state: UIGestureRecognizerState = .possible {
         didSet {
-            if state == oldValue { return }
+            if
+                state == oldValue || oldValue == .cancelled && state != .ended
+            {
+                return
+            }
+
             onStateChanged?()
+
             switch state {
-            case .failed, .cancelled:
-                // touchesCancelled(touches: Set<UITouch>, with: UIEvent)
+            case .recognized, .ended, .failed:
                 state = .possible
-            case .recognized, .ended:
-                state = .possible
-            case .changed:
+            case .began, .changed:
```

Comment:
i don't think it's correct to cancel gesture recognisers on `.began` –  `.began` is an "open" starting state that _might_ lead to something happening but is not yet certain – that's why we used to wait until `.changed` here. I think all of this stuff needs code comments though because it's pretty complex and often unclear why things are the way they are

Reply:
but how would that work for TapGestureRecognizers, do they emit `.changed` at all?

Reply:
I actually don't know – I would have expected they should, yes

Reply:
thats basically the change I applied to the view lifecycles (recognizer vs view). they don't get called if there is a tap recognizer which has began or a panRecognizer which has moved. This change is not properly reflected in this code here (recognizer vs recognizer). I need to wrap my head around this again.

Reply:
Ok I am getting the feeling more and more that part of the fix would be ensuring that TapGestureRecognizers get set to the `.changed` state at some point in their lifecycle in order to make things like this work as expected? Would be good to understand what iOS does here